### PR TITLE
feat: support Tencent Cloud Log Service 

### PR DESCRIPTION
**Pull request by ychensha**
### Description
[product doc](https://cloud.tencent.com/document/product/614)

apisix log util is more powerfull than nginx access log, and it's useful for debugging and replaying later.

Fixes # (7592)

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by soulbird** [comment]
Thanks for your contribution. But please add a license to each file and add test cases first.

**Comment by ychensha** [comment]
> 

any doc for testing? I have no exp for env setup & PERL testing framework...

**Comment by soulbird** [comment]
> > 
> 
> any doc for testing? I have no exp for env setup & PERL testing framework...

See: https://github.com/apache/apisix/blob/master/docs/en/latest/internal/testing-framework.md

**Comment by ychensha** [comment]
> > > 
> > 
> > 
> > any doc for testing? I have no exp for env setup & PERL testing framework...
> 
> See: https://github.com/apache/apisix/blob/master/docs/en/latest/internal/testing-framework.md

tried..  I work on Mac, and now trapped in etcd configuration for more tesing case

File: apisix/plugins/tencent-cloud-cls/cls-sdk.lua [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+    -- TODO: support lz4/zstd compress
+    params_cache.method = "POST"
+    params_cache.body = pb_data
+
+    local cls_url = "http://" .. host .. cls_api_path .. "?topic_id=" .. topic
+    core.log.debug("CLS request URL: ", cls_url)
+
+    local res, err = http_new:request_uri(cls_url, params_cache)
+    if not res then
+        return false, err
```
Comment (spacewander) [inline apisix/plugins/tencent-cloud-cls/cls-sdk.lua:130]: Would be better to close the http client. What about introducing a new function?

```
client = xxx
local ok, err = do_something()
client:close()
```

File: apisix/plugins/tencent-cloud-cls/cls-sdk.lua [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+            contents = contents,
+        })
+        :: continue ::
+    end
+
+    insert_tab(log_group_list, {
+        logs = format_logs,
+        source = host_ip,
+    })
+    local data = assert(pb.encode("cls.LogGroupList", log_group_list_pb))
```
Comment (spacewander) [inline apisix/plugins/tencent-cloud-cls/cls-sdk.lua:212]: https://github.com/apache/apisix/blob/18a6cacf1681c80d7821b0791f6b704fe6a9a4d8/apisix/core/pubsub.lua#L76
pb.encode will throw an error when failed

File: apisix/plugins/tencent-cloud-cls/cls-sdk.lua [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+
+local pb = require "pb"
+local assert = assert
+assert(pb.loadfile("apisix/plugins/tencent-cloud-cls/cls.pb"))
```
Comment (spacewander) [inline apisix/plugins/tencent-cloud-cls/cls-sdk.lua:20]: https://github.com/apache/apisix/blob/18a6cacf1681c80d7821b0791f6b704fe6a9a4d8/apisix/core/pubsub.lua#L50
loadfile will throw an error when failed

File: apisix/plugins/tencent-cloud-cls/cls-sdk.lua [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    for _, v in ipairs(resolved.ip) do
+        insert_tab(ListTab, v)
+    end
+    return ListTab
+end
+
+local host_ip = tostring(unpack(get_ip(core_gethostname())))
+local log_group_list = {}
+local log_group_list_pb = {
+    logGroupList = log_group_list,
```
Comment (spacewander) [inline apisix/plugins/tencent-cloud-cls/cls-sdk.lua:73]: Let's use underscore style

File: apisix/plugins/tencent-cloud-cls/cls-sdk.lua [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+        total_size = total_size + log_size
+        if total_size > MAX_LOG_GROUP_VALUE_SIZE then
+            insert_tab(log_group_list, {
+                logs = format_logs,
+                source = host_ip,
+            })
+            format_logs = new_tab(#logs - i, 0)
+            total_size = 0
+            local data = assert(pb.encode("cls.LogGroupList", log_group_list_pb))
+            send_cls_request(host, topic_id, secret_id, secret_key, data)
```
Comment (spacewander) [inline apisix/plugins/tencent-cloud-cls/cls-sdk.lua:198]: Not error handling?

File: apisix/plugins/tencent-cloud-cls/cls-sdk.lua (hunk 3 above)
Comment (spacewander) [reply apisix/plugins/tencent-cloud-cls/cls-sdk.lua:20]: And we can load the proto file directly? So we don't need to store another .pb file.

File: apisix/plugins/tencent-cloud-cls/cls.proto [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// https://cloud.tencent.com/document/product/614/59470
+package cls;
```
Comment (spacewander) [inline apisix/plugins/tencent-cloud-cls/cls.proto:19]: Let's update https://github.com/apache/apisix/blob/18a6cacf1681c80d7821b0791f6b704fe6a9a4d8/Makefile#L251 to install the proto file

File: apisix/plugins/tencent-cloud-cls/cls.proto [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+message Log
+{
+  message Content
+  {
+    required string key   = 1; // 每组字段的 key
```
Comment (spacewander) [inline apisix/plugins/tencent-cloud-cls/cls.proto:25]: Please use English

File: apisix/plugins/tencent-cloud-cls.lua [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+function _M.log(conf, ctx)
+    -- sample if set
+    if ctx.cls_sample == nil then
+        core.log.debug("not sampled")
+        return
+    end
+    local entry = log_util.get_full_log(ngx, conf)
+    if not entry.route_id then
+        entry.route_id = "no-matched"
```
[the remark this answers is not in the document]
Comment (ychensha) [reply apisix/plugins/tencent-cloud-cls.lua:114]: No need. APISIX log is easy for debugging, if route_id field is missing, we known what happened

File: apisix/plugins/tencent-cloud-cls.lua [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    name = plugin_name,
+    schema = batch_processor_manager:wrap_schema(schema),
+}
+
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
+end
+
+function _M.body_filter(conf, ctx)
+    -- sample if set
```
[the remark this answers is not in the document]
Comment (ychensha) [reply apisix/plugins/tencent-cloud-cls.lua:78]: OK.  I will do in that way..

File: apisix/plugins/tencent-cloud-cls/cls-sdk.lua (hunk 5 above)
Comment (ychensha) [reply apisix/plugins/tencent-cloud-cls/cls-sdk.lua:203]: Any suggestion? Retry in batch processor is hard to understand here...

File: apisix/plugins/tencent-cloud-cls/cls-sdk.lua (hunk 4 above)
Comment (ychensha) [reply apisix/plugins/tencent-cloud-cls/cls-sdk.lua:72]: It is pb field

File: apisix/plugins/tencent-cloud-cls/cls-sdk.lua (hunk 3 above)
Comment (ychensha) [reply apisix/plugins/tencent-cloud-cls/cls-sdk.lua:20]: I will drop .proto file too...  `protoc.load` like otel lib

File: apisix/plugins/tencent-cloud-cls.lua [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+
+
+local plugin_name = "tencent-cloud-cls"
+local batch_processor_manager = bp_manager_mod.new(plugin_name)
+local schema = {
+    type = "object",
+    properties = {
+        cls_host = { type = "string" },
+        cls_topic = { type = "string" },
+        -- https://console.cloud.tencent.com/capi
```
Comment (tzssangglass) [inline apisix/plugins/tencent-cloud-cls.lua:35]: what is this link for?

File: apisix/plugins/tencent-cloud-cls.lua [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+
+function _M.access(conf, ctx)
+    -- sample if set
+    ctx.cls_sample = false
+    if conf.sample_ratio == 1 or math.random() < conf.sample_ratio then
+        core.log.debug("cls sampled")
+        ctx.cls_sample = true
+        return
+    end
+    core.log.debug("cls not sampled")
```
Comment (tzssangglass) [inline apisix/plugins/tencent-cloud-cls.lua:90]: No meaningful log?

**Comment by tzssangglass** [comment]
I have a question: Is it that we can't do the test case to verify that this plugin is working correctly?
Because it requires a Tencent account and to make sure that this account is always available.

File: apisix/plugins/tencent-cloud-cls/cls-sdk.lua (hunk 5 above)
Comment (spacewander) [reply apisix/plugins/tencent-cloud-cls/cls-sdk.lua]: We can simply log down the err?

**Comment by spacewander** [comment]
> I have a question: Is it that we can't do the test case to verify that this plugin is working correctly? Because it requires a Tencent account and to make sure that this account is always available.

We can inject a hook or mock into the code? Like:
https://github.com/apache/apisix/blob/b3d4b464045b9bdfed3170cbd7a343ed5c40b24a/t/plugin/zipkin2.t#L35
https://github.com/apache/apisix/blob/b3d4b464045b9bdfed3170cbd7a343ed5c40b24a/t/plugin/opentelemetry.t#L553

**Comment by ychensha** [comment]
> 

I have same question too. But same question can be applie to other plugins that have env matters

**Comment by ychensha** [comment]
> > 
> 
> I have same question too. But same question can be applie to other plugins that have env matters

**Comment by ychensha** [comment]
@spacewander @tzssangglass after code related conversations clsoed, I will do the function test again for the related code modification.

**Comment by tzssangglass** [comment]
> We can inject a hook or mock into the code? Like:

@ychensha Let's do that for now, there's no better way to do it.

**Comment by ychensha** [comment]
> 

If hook is ok for now, I will add test code like otel plugin.

**Comment by spacewander** [comment]
> > 
> 
> If hook is ok for now, I will add test code like otel plugin.

Don't be afraid. Feel free to contact me once you have trouble with the hook.

**Comment by spacewander** [comment]
To fix https://github.com/apache/apisix/runs/7851928549?check_suite_focus=true
Let's add tencent-cloud-cls to https://github.com/apache/apisix/blob/b0934f9ed45ec352c3de09f59e6912a7ba71936e/Makefile#L344

**Comment by spacewander** [comment]
https://github.com/apache/apisix/runs/7861163309?check_suite_focus=true
Let's add the `tencent-cloud-cls`

File: apisix/plugins/tencent-cloud-cls/cls-sdk.lua [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+local pairs = pairs
+local type = type
+local tostring = tostring
+local setmetatable = setmetatable
+local pcall = pcall
+
+local MAX_SINGLE_VALUE_SIZE = 1 * 1024 * 1024
+local MAX_LOG_GROUP_VALUE_SIZE = 5 * 1024 * 1024 -- 5MB
+
+local cls_api_path = "/structuredlog"
```
Comment (soulbird) [inline apisix/plugins/tencent-cloud-cls/cls-sdk.lua:48]: Is this value immutable?

File: apisix/plugins/tencent-cloud-cls/cls-sdk.lua (hunk 1 above)
Comment (spacewander) [reply apisix/plugins/tencent-cloud-cls/cls-sdk.lua:140]: @ychensha 
Please address this.

File: t/plugin/tencent-cloud-cls.t [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+
+
+
+=== TEST 4: access local server
+--- request
+GET /opentracing
+--- response_body
+opentracing
+--- error_log
+Batch Processor[tencent-cloud-cls] successfully processed the entries
```
Comment (spacewander) [inline t/plugin/tencent-cloud-cls.t:154]: We need to also verify the log data received by the stub server or via injected check.

File: t/plugin/tencent-cloud-cls.t [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+--- config
+    location /t {
+        content_by_lua_block {
+            local t = require("lib.test_admin").test
+            local code, body = t('/apisix/admin/routes/1',
+                 ngx.HTTP_PUT,
+                 [[{
+                        "plugins": {
+                            "tencent-cloud-cls": {
+                                "cls_host": "127.0.0.1:10420",
```
Comment (spacewander) [inline t/plugin/tencent-cloud-cls.t:116]: Let's add a test that  the log data is sent to incorrect address

File: t/plugin/tencent-cloud-cls.t [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+
+
+=== TEST 4: access local server
+--- request
+GET /opentracing
+--- response_body
+opentracing
+--- error_log
+Batch Processor[tencent-cloud-cls] successfully processed the entries
+--- wait: 0.5
```
Comment (spacewander) [inline t/plugin/tencent-cloud-cls.t:155]: Let's add a test that we check the log data with the custom log format from plugin metadata.

File: apisix/plugins/tencent-cloud-cls/cls-sdk.lua [hunk 16]
```diff
@@ -0,0 +1,308 @@
...
+
+local _M = { version = 0.1 }
+local mt = { __index = _M }
+
+local pb_state
+local function init_pb_state()
+    pb.state(nil)
+    protoc.reload()
+    local cls_sdk_protoc = protoc.new()
+    if not cls_sdk_protoc.loaded["tencent-cloud-cls/cls.proto"] then
```
Comment (spacewander) [inline apisix/plugins/tencent-cloud-cls/cls-sdk.lua:196]: we can skip this check like https://github.com/apache/apisix/pull/7686

File: apisix/plugins/tencent-cloud-cls/cls-sdk.lua [hunk 17]
```diff
@@ -0,0 +1,308 @@
...
+local insert_tab = table.insert
+local ipairs = ipairs
+local pairs = pairs
+local type = type
+local tostring = tostring
+local setmetatable = setmetatable
+local pcall = pcall
+
+local MAX_SINGLE_VALUE_SIZE = 1 * 1024 * 1024
+local MAX_LOG_GROUP_VALUE_SIZE = 5 * 1024 * 1024 -- 5MB
```
Comment (spacewander) [inline apisix/plugins/tencent-cloud-cls/cls-sdk.lua:46]: Could you give us the source of these two magic number? Thanks!

File: Makefile [hunk 18]
```diff
@@ -345,6 +345,9 @@ install: runtime
 	$(ENV_INSTALL) -d $(ENV_INST_LUADIR)/apisix/plugins/zipkin
 	$(ENV_INSTALL) apisix/plugins/zipkin/*.lua $(ENV_INST_LUADIR)/apisix/plugins/zipkin/
 
+	$(ENV_INSTALL) -d $(ENV_INST_LUADIR)/apisix/plugins/tencent-cloud-cls
```
Comment (spacewander) [inline Makefile:348]: Let's put it in alphabetical order https://github.com/apache/apisix/pull/7698

File: apisix/plugins/tencent-cloud-cls/cls-sdk.lua (hunk 17 above)
Comment (spacewander) [reply apisix/plugins/tencent-cloud-cls/cls-sdk.lua:46]: @ychensha 
Let's add a comment for the source

File: apisix/plugins/tencent-cloud-cls/cls-sdk.lua [hunk 19]
```diff
@@ -0,0 +1,308 @@
...
+    return normalized_log, log_size
+end
+
+
+local _M = { version = 0.1 }
+local mt = { __index = _M }
+
+local pb_state
+local function init_pb_state()
+    pb.state(nil)
```
Comment (spacewander) [inline apisix/plugins/tencent-cloud-cls/cls-sdk.lua:193]: Let's improve this part like https://github.com/apache/apisix/pull/7728

File: apisix/plugins/tencent-cloud-cls/cls-sdk.lua [hunk 20]
```diff
@@ -0,0 +1,308 @@
...
+    return setmetatable(self, mt)
+end
+
+
+function _M.send_to_cls(self, logs)
+    clear_tab(log_group_list)
+    local now = ngx_now() * 1000
+
+    -- recovery of stored pb_store
+    pb.state(pb_state)
```
Comment (spacewander) [inline apisix/plugins/tencent-cloud-cls/cls-sdk.lua:263]: Better to store / restore state in the function we used it (send_cls_request in this case)

**Review by spacewander — COMMENTED** [review]
Some comments need to be handled:
https://github.com/apache/apisix/pull/7593/files#r947501015
https://github.com/apache/apisix/pull/7593/files#r937476352
https://github.com/apache/apisix/pull/7593/files#r947500216

**Comment by tzssangglass** [comment]
@ychensha pls don't use force-push to update code, or reviews' comments may be ignore.

File: t/plugin/tencent-cloud-cls.t [hunk 21]
```diff
@@ -0,0 +1,311 @@
...
+--- extra_init_by_lua
+    local cls = require("apisix.plugins.tencent-cloud-cls.cls-sdk")
+    cls.send_cls_request = function(self, pb_obj)
+        if (#pb_obj.logGroupList ~= 1) then
+            ngx.log(ngx.ERR, "unexpected logGroupList length: ", #pb_obj.logGroupList)
+            return false
+        end
+        local log_group = pb_obj.logGroupList[1]
+        if #log_group.logs ~= 1 then
+            ngx.log(ngx.ERR, "unexpected logs length: ", #log_group.logs)
```
Comment (spacewander) [inline t/plugin/tencent-cloud-cls.t:257]: Let's use no_error_log to ensure the error log doesn't exist.

File: t/plugin/tencent-cloud-cls.t [hunk 22]
```diff
@@ -0,0 +1,311 @@
...
+            return false
+        end
+        local log = log_group.logs[1]
+        for k, v in ipairs(log.contents) do
+            ngx.log(ngx.ERR, "key:", v.key, ",value:", v.value)
```
Comment (spacewander) [inline t/plugin/tencent-cloud-cls.t:262]: It seems we doesn't check it in the error log?

File: t/plugin/tencent-cloud-cls.t [hunk 23]
```diff
@@ -0,0 +1,311 @@
...
+        listen 10421;
+        location /structuredlog {
+            content_by_lua_block {
+                ngx.req.read_body()
+                local data = ngx.req.get_body_data()
+                local headers = ngx.req.get_headers()
+                ngx.log(ngx.WARN, "tencent-cloud-cls body: ", data)
+                for k, v in pairs(headers) do
+                    ngx.log(ngx.WARN, "tencent-cloud-cls headers: " .. k .. ":" .. v)
+                end
```
Comment (spacewander) [inline t/plugin/tencent-cloud-cls.t:61]: We can simply use `ngx.exit(500)` as we don't care about the input.

File: t/plugin/tencent-cloud-cls.t [hunk 24]
```diff
@@ -0,0 +1,311 @@
...
+
+
+
+=== TEST 10: log use log_format
+--- request
+GET /opentracing
+--- response_body
+opentracing
+--- error_log
+"@timestamp":[true,"time_iso8601"]
```
Comment (spacewander) [inline t/plugin/tencent-cloud-cls.t:310]: Where is this error_log from? Maybe we can use a longer log message so people can know its position.

File: apisix/plugins/tencent-cloud-cls.lua [hunk 25]
```diff
@@ -0,0 +1,144 @@
...
+
+
+local plugin_name = "tencent-cloud-cls"
+local batch_processor_manager = bp_manager_mod.new(plugin_name)
+local schema = {
+    type = "object",
+    properties = {
+        cls_host = { type = "string" },
+        cls_topic = { type = "string" },
+        -- https://console.cloud.tencent.com/capi
```
Comment (tzssangglass) [inline apisix/plugins/tencent-cloud-cls.lua:35]: this API needs to do auth and have nothing to show directly, remove it or change another link.

**Review by tzssangglass — CHANGES_REQUESTED** [review]

File: apisix/plugins/tencent-cloud-cls.lua (hunk 11 above)
Comment (tzssangglass) [reply apisix/plugins/tencent-cloud-cls.lua:90]: ping @ychensha you need to respond to the review comments, not just close them.

File: apisix/plugins/tencent-cloud-cls.lua [hunk 26]
```diff
@@ -0,0 +1,144 @@
...
+
+    local ok, err = core.schema.check(schema, conf)
+    if not ok then
+        return nil, err
+    end
+    return log_util.check_log_schema(conf)
+end
+
+
+function _M.access(conf, ctx)
```
Comment (tzssangglass) [inline apisix/plugins/tencent-cloud-cls.lua:82]: we can do this in the `rewrite` phase?

Comment (tzssangglass) [reply apisix/plugins/tencent-cloud-cls.lua:82]: my mistake, keep `access` phase.

File: apisix/plugins/tencent-cloud-cls/cls-sdk.lua [hunk 27]
```diff
@@ -0,0 +1,315 @@
...
+    client:close()
+    return res, err
+end
+
+
+function _M.send_cls_request(self, pb_obj)
+    -- recovery of stored pb_store
+    pb.state(pb_state)
+
+    local ok, pb_data = pcall(pb.encode, "cls.LogGroupList", pb_obj)
```
Comment (spacewander) [inline apisix/plugins/tencent-cloud-cls/cls-sdk.lua:227]: Need to recover the pb_state after used.

File: apisix/plugins/tencent-cloud-cls/cls-sdk.lua [hunk 28]
```diff
@@ -0,0 +1,315 @@
...
+    if not ok or not pb_data then
+        core.log.error("failed to encode LogGroupList, err: ", pb_data)
+        return false, pb_data
+    end
+
+    local client = http:new()
+    client:set_timeouts(cls_conn_timeout, cls_send_timeout, cls_read_timeout)
```
Comment (spacewander) [inline apisix/plugins/tencent-cloud-cls/cls-sdk.lua:234]: The client is unused now?

File: t/plugin/tencent-cloud-cls.t [hunk 29]
```diff
@@ -0,0 +1,304 @@
...
+            return false
+        end
+        return true
+    end
+--- request
+GET /opentracing
+--- response_body
+opentracing
+--- no_error_log
+[error]
```
Comment (spacewander) [inline t/plugin/tencent-cloud-cls.t:265]: We don't need to set `no_error_log` again as we already set at the top of this file:
```
if ((!defined $block->error_log) && (!defined $block->no_error_log)) {
        $block->set_value("no_error_log", "[error]");
    }
```

File: t/plugin/tencent-cloud-cls.t [hunk 30]
```diff
@@ -0,0 +1,304 @@
...
+                )
+            ngx.say(body)
+        }
+    }
+--- request
+GET /t
+--- response_body
+passed
+--- no_error_log
+[error]
```
Comment (spacewander) [inline t/plugin/tencent-cloud-cls.t:293]: Ditto.
And also the `--- request` can be saved.

File: t/plugin/tencent-cloud-cls.t [hunk 31]
```diff
@@ -0,0 +1,304 @@
...
+
+
+
+=== TEST 10: log use log_format
```
Comment (spacewander) [inline t/plugin/tencent-cloud-cls.t:297]: Could we also check the log content like `=== TEST 8: verify cls api request`?

**Comment by spacewander** [comment]
Let's merge master to make CI pass.

**Review by spacewander — APPROVED** [review]
Approve except https://github.com/apache/apisix/pull/7593/files#r941961486
As this PR is big enough, let's open another PR for the doc.

**Review by spacewander — APPROVED** [review]

**Comment by spacewander** [comment]
As this PR is big enough, let's open another PR for the doc.

**Review by tzssangglass — APPROVED** [review]

**Pull request merged**